Port AST to Rust

krobelus opened the pull request:
The translation is fairly direct though it adds some duplication, for example there are multiple "match" statements that mimic function overloading.

Rust has no overloading, and we cannot have generic methods in the Node trait (due to a Rust limitation, the error is like "cannot be made into an object") so we include the type name in method names.

Give clients like "indent_visitor_t" a Rust companion ("IndentVisitor") that takes care of the AST traversal while the AST consumption remains in C++ for now.  In future, "IndentVisitor" should absorb the entirety of "indent_visitor_t".  This pattern requires that "fish_indent" be exposed includable header to the CXX bridge.

Alternatively, we could define FFI wrappers for recursive AST traversal. 
Rust requires we separate the AST visitors for "mut" and "const" scenarios. Take this opportunity to concretize both visitors:

The only client that requires mutable access is the populator.  To match the structure of the C++ populator which makes heavy use of function overloading, we need to add a bunch of functions to the trait. Since there is no other mutable visit, this seems acceptable.

The "const" visitors never use "will_visit_fields_of()" or "did_visit_fields_of()", so remove them (though this is debatable).

Like in the C++ implementation, the AST nodes themselves are largely defined via macros.  Union fields like "Statement" and "ArgumentOrRedirection" do currently not use macros but may in future.

In addition to node-specific children, each node has fields "children", "range", "parse_token_type" and "keyword" which are zero-size unless the node is a "List", "Leaf", "Token" or "Keyword". Defining this many unused fields is a bit ugly but it's not so bad because they are private.

This commit also introduces a precedent for a type that is defined in one CXX bridge and used in another one - "ParseErrorList". To make this work we need to manually define "ExternType".

There is one annoyance with CXX: functions that take explicit lifetime parameters require to be marked as unsafe. This makes little sense because functions that return `&Foo` with implicit lifetime can be misused the same way on the C++ side.

One notable change is that we cannot directly port "find_block_open_keyword()" - which is used to compute an error - because it relies on the stack of visited nodes. We cannot modify a stack of node references while we do the "mut" walk. Happily, a more idiomatic solution is easy: we can tell the AST visitor to backtrack to the parent node and create the error there.

Since "node_t::accept_base" is no longer a template we don't need the "node_visitation_t" trampoline anymore.

Leftovers:
- Enum variants are still snakecase; I didn't get around to changing this yet.
- "ast_type_to_string()" still returns a snakecase name. This could be changed, it's not user visible.

Xiretza reviewed (commented):
A couple comments on the `unescape_string` port, haven't had time to look at the actual AST changes yet.

mqudsi commented on fish-rust/src/common.rs:
```diff
@@ -1,11 +1,62 @@
+// Make sure the ranges defined above don't exceed the range for non-characters.
+// This is to make sure we didn't do something stupid in subdividing the
+// Unicode range for our needs.
+// #if WILDCARD_RESERVED_END > RESERVED_CHAR_END
+// #error
+// #endif
```
The other option is to use a `#[repr(u32)]` enum and do `Xxx = 'x' as u32,` for each entry, then impl `Into<char> for Enum`. Not an ideal solution, but it would catch any cases of accidentally assigning the same value to two different entries and allow (should we need it) exhaustiveness checking.

krobelus replied:
thanks for the feedback, I've added it to the port of all of common.h in https://github.com/fish-shell/fish-shell/pull/9688

mqudsi commented on fish-rust/src/common.rs:
```diff
@@ -62,6 +113,17 @@ pub struct EscapeFlags {
     pub symbolic: bool,
 }
 
+/// Flags for unescape_string functions.
+pub type UnescapeFlags = u32;
+/// default behavior
+pub const UNESCAPE_DEFAULT: UnescapeFlags = 0;
+/// escape special fish syntax characters like the semicolon
+pub const UNESCAPE_SPECIAL: UnescapeFlags = 1 << 0;
+/// allow incomplete escape sequences
+pub const UNESCAPE_INCOMPLETE: UnescapeFlags = 1 << 1;
+/// don't handle backslash escapes
+pub const UNESCAPE_NO_BACKSLASHES: UnescapeFlags = 1 << 2;
```
Wouldn't the most ideal solution be to define these as `bitflags!`? We already take a (transitive) dependency on that and it's a very well-supported and well-known crate.

Xiretza replied:
Probably depends on your definition of "most ideal", it uses a couple bytes less memory but also makes the code slightly more opaque for human readers.

krobelus replied:
If we use a struct we'll want to come up with a nice way of creating it when only few bits are set (c.f. https://github.com/fish-shell/fish-shell/pull/9588). Builder pattern?
I like bitflags. It's a bit low-level - we really only need to specify the representation if/while we interface with C++ but that's okay.

We have dozens of calls to `escape_string` that will be ported incrementally.
We should make porting easy. To minimize surprise, we should use the same interface for the C++ and Rust version of a function. I'm inclined to keep bit fields for now and maybe revisit this after the port is done. Unless there is an obvious and nice-to-use equivalent (including how to call it etc).
The bit fields are a fairly isolated component, it won't hold anything else back.

Also, how do we model the default argument to `escape_string`?

krobelus replied:
yeah I'd go with `bitflags!`. I've manually implemented BitAnd and friends because I wasn't aware that we already use it

mqudsi commented on src/ffi_baggage.h:
```diff
@@ -0,0 +1,9 @@
+// Symbols that get autocxx bindings but are not used in a given binary, will cause "undefined
+// reference" when trying to link that binary. Work around this by marking them as used in
+// all binaries.
+void mark_as_used() {
+    //
+    pretty_printer_t({}, {});
+}
```
I think `__attribute__ ((used))` on the constructor [should accomplish the same](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes:~:text=for%20this%20function.-,used,-This%20attribute%2C%20attached)? 

> This attribute, attached to a function, means that code must be emitted for the function even if it appears that the function is not referenced. This is useful, for example, when the function is referenced only in inline assembly.
> 
> When applied to a member function of a C++ class template, the attribute also means that the function is instantiated if the class itself is instantiated.

krobelus replied:
hmm `__attribute__((used))` does not seem to work

mqudsi commented on src/reader.cpp:
```diff
@@ -4737,16 +4738,16 @@ static int read_ni(parser_t &parser, int fd, const io_chain_t &io) {
     auto errors = new_parse_error_list();
-    auto ast = ast::ast_t::parse(str, parse_flag_none, &*errors);
-    bool errored = ast.errored();
+    auto ast = ast_parse(str, parse_flag_none, &*errors);
```
In general to avoid lots of cascading changes to the cpp code when replacing a type `T` with a `unique_ptr<T>` returned from rust, you can just deconstruct the unique_ptr at the call site in a one-liner, e.g.

```cpp
auto ast = std::move(*ast_parse(...));
```

but when it's a box you would need to shadow the declaration since `rust::box<T>` only lets you acquire the raw pointer and not the value, so you'd need

```cpp
auto shadow_ast = ast_parse(...);
auto &ast = *shadow_ast;
```

then the downstream code won't need to be changed.

mqudsi commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+
+    // Cast to any sub-trait.
+
+    fn as_keyword(&self) -> Option<&dyn Keyword> {
+        None
+    }
+    fn as_leaf(&self) -> Option<&dyn Leaf> {
+        None
+    }
+    fn as_token(&self) -> Option<&dyn Token> {
+        None
+    }
```
Depending on how these are used, you might have another option. You can (and this is the idiomatic rust way where possible) avoid returning an `Option<T>` at all by implementing these trait methods (here) only for types that implement the sub-trait by just putting a generic constraint on the default method impl (but then you can't just blindly call them on any random node to introspect whether or not it implements the trait if you're doing it all via dynamic dispatch and never via a concrete type).

mqudsi commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+// pub struct NodeFFI<'a>(pub Option<&'a dyn Node>);
+
+#[derive(Clone)]
+pub enum NodeFFI<'a> {
```
Minor nit: Have we settled on all-caps FFI? Rust generally uses `NodeFfi` as the naming convention even for abbreviations.

mqudsi commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+    }
+}
+
+impl String_ {
```
We could rename this to Text and use `[cxx_name]` to alias it as string.

mqudsi commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+    }
+}
+
+impl AndorJobList {
```
We might as well correct the capitalization and mkae this `AndOrXxx` everywhere.

mqudsi reviewed (commented):
Awesome work, this is a monster of a PR. GitHub is struggling to load the diffs and everything is very laggy typing out comments!

ridiculousfish commented:
This is a huge chunk of work, very impressive. I've got some pending review comments but haven't finished reviewing, hopefully tomorrow!

ridiculousfish commented:
I added some comments but haven't touched the AST parts itself. "Tomorrow" was way too optimistic; this is a massive PR and I will get to the AST portion.

ridiculousfish reviewed (changes requested):
I honestly struggled to review this and got pretty lost. I think it is more complex than necessary, and the long list of methods like `as_mut_while_header`, and desire to avoid `unsafe`, is part of the source of the complexity. I think we should try it with a design closer to the C++ one, even if that results in more unsafe:

- Have a bunch of structs like JobPipeline, and a parallel enum or set of constants (which you have as `Type`).
- A trait which exposes the Type for each struct via an associated constant
- A method in the trait to convert to a concrete type which is given as a generic.

Example:

```
trait Node {
    const TYPE: Type;
	
	fn get_type(&self) -> Type;
	
	fn try_as<Target: Node>(&self) -> Option<&Target> {
	    if self.get_type() == Target::TYPE {
		    let ref = unsafe { &*(self as *const Target) };
		    Some(ref)
		} else {
		   None
		}
	}
}
```

and then the caller can write:

```
let s: Option<&Statement> = node.try_as();
```

eliminating a large amount of methods.

Another source of complexity is the desire to handle all the node types uniformly, in particular with that single `define_nodes!` mega-invocation. I think it would be better to handle Branches, Leaves, and Lists separately. We should work hard to make the struct declarations ordinary Rust code. We should be able to write something like this:

```
struct BlockStatement {
    // A header like for, while, etc.
    header: Header,

    // List of jobs in this block.
    jobs: JobList,

    // The 'end' node.
    end: Keyword<KeywordType::End>,

    // Arguments and redirections associated with the block.
    args_or_redirs: ArgumentOrRedirectionList,
};
make_branch_node!(
	BlockStatement,
	Type::BlockStatement,
	
	header, jobs, end, args_or_redirs
);

By handling branches, leaves, and lists separately I think each individual macro will be smaller.

Xiretza commented:
> Example:
> 
> ```rust
> trait Node {
> 	const TYPE: Type;
> 	
> 	fn get_type(&self) -> Type;
> 	
> 	fn try_as<Target: Node>(&self) -> Option<&Target> {
> 	    if self.get_type() == Target::TYPE {
> 		    let ref = unsafe { &*(self as *const Target) };
> 		    Some(ref)
> 		} else {
> 		   None
> 		}
> 	}
> }
> ```

Note that this is the exact signature of [`Any::downcast_ref()`](https://doc.rust-lang.org/std/any/trait.Any.html#method.downcast_ref) - maybe `Node` can just get `Any` as a supertrait to use the existing dynamic typing infrastructure?

ridiculousfish commented:
I don't see the advantage in using `Any`. Nodes should be convertible to concrete Node types, not Any type.

Xiretza commented:
The advantage is that `Any`'s type ID and conversion machinery doesn't have to be reimplemented.

Any concrete type will only ever be downcastable to itself anyway, so I don't really understand the issue. You could always add a simple method on `Node` that restricts the target type to another `Node` type and just calls `downcast_ref()`, if you want the constraint to be reflected in the type signature.

krobelus commented:
I did not yet figure out how to put generic methods like `try_as` into the `Node` trait.
This is the error I get

<details>

```diff
diff --git a/fish-rust/src/ast.rs b/fish-rust/src/ast.rs
index 50e1fa3f2..c37dfebe7 100644
--- a/fish-rust/src/ast.rs
+++ b/fish-rust/src/ast.rs
@@ -113,6 +113,18 @@ impl<T: AcceptorMut> AcceptorMut for Option<T> {
 
 /// Node is the base trait of all AST nodes.
 pub trait Node: Acceptor + AcceptorMut + std::fmt::Debug {
+    const TYPE: Type;
+
+    fn try_as<Target: Node>(&self) -> Option<&Target> {
+        if self.typ() == Target::TYPE {
+            // let ref = unsafe { &*(self as *const Target) };
+            // Some(ref)
+            None
+        } else {
+            None
+        }
+    }
+
     /// The parent node, or null if this is root.
     fn parent(&self) -> Option<&dyn Node>;
     fn parent_ffi(&self) -> &Option<*const dyn Node>;
@@ -649,6 +661,8 @@ macro_rules! define_node {
             }
         }
         impl Node for $name {
+            const TYPE: Type = Type::$type;
+
             trait_downcasts!($name, $category $(, $base)*);
             downcasts!($name $(, $conversion, $conversion_mut)*);
```

</details>

```
error[E0038]: the trait `ast::Node` cannot be made into an object
   --> src/highlight.rs:79:35
    |
79  |     fn visit(&mut self, node: &'a dyn Node) {
    |                                   ^^^^^^^^ `ast::Node` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> src/ast.rs:116:11
    |
115 | pub trait Node: Acceptor + AcceptorMut + std::fmt::Debug {
    |           ---- this trait cannot be made into an object...
116 |     const TYPE: Type;
    |           ^^^^ ...because it contains this associated `const`
117 |
118 |     fn try_as<Target: Node>(&self) -> Option<&Target> {
    |        ^^^^^^ ...because method `try_as` has generic type parameters
    = help: consider moving `try_as` to another trait
    = help: consider moving `TYPE` to another trait
```

krobelus commented:
we could make `Node` an enum instead of a trait

krobelus commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+        None
+    }
+
+    fn as_mut_redirection(&mut self) -> Option<&mut Redirection> {
```
Good point, I missed that we can make a private trait that extends the public one.

Xiretza commented:
Looks like leveraging `Any` won't work due to lack of [dyn upcasting coercion](https://github.com/rust-lang/rust/issues/65991) - can't go from a `&dyn Node` to a `&dyn Any`, even if `Node: Any`.

krobelus commented:
> We should work hard to make the struct declarations ordinary Rust code.

I'll see what I can do but if we want to keep embedding fields like `parent` and `source_range`, then defining the struct outside the macro means either
- adding fields for `parent` etc to every struct definition
- using a proc macro. I made everything work without proc macros, hoping that this will force the implementation to remain comprehensible. Maybe it's time to reconsider.

>     end: Keyword<KeywordType::End>,

I did not implement Keyword as a generic struct because there are keyword nodes that allow multiple keywords and Rust does not have variadic generics. However I realized we can probably pass a slice, like `struct<T: &'static [KeywordType]> Keyword {...}`

> GitHub is struggling to load the diffs and everything is very laggy typing out comments!

@mqudsi one workaround is https://github.com/danobi/prr

krobelus commented:
I used to use `Any` by implementing `as_any()` on every node but that's not better than the current solution.

krobelus commented:
I guess we can try to use generics if we give up dynamic dispatching (so no `dyn Node`), which means we'd need to use generics  in the `visit` methods.

mqudsi commented:
Depending on how much of a difference there is between the size of the smallest, largest, and most common `Node` types, using an enum instead could work. Even if there's a big size difference, you can work around that by boxing the larger variants (cheaper than all the dyn casting anyway).

We don't expect Node to be extended from outside the module or from other crates/projects so being locked into a set number of variants isn't a problem. It would make the code as "regular rust" as possible and make a lot of things cheaper. You'd still need to box nodes in order to get around the limitations on self-referential structs, but it would be more straight-forward and give the compiler more to work with when it comes to optimizing.

ridiculousfish commented:
`dyn Node` should be fine; we don't need `Any`. Maybe this is just my experiences with C++ RTTI talking but I don't fully trust Any.

ridiculousfish commented:
Shared properties such as `source_range` and `parent` may be combined into a single type.

```rust
struct Base {
   source_range: SourceRange,
   parent: Option<*const dyn Node>,
};

struct BlockStatement {
    base: Base,

    // A header like for, while, etc.
    header: Header,

    // List of jobs in this block.
    jobs: JobList,

    // The 'end' node.
    end: Keyword<KeywordType::End>,

    // Arguments and redirections associated with the block.
    args_or_redirs: ArgumentOrRedirectionList,
}
```

We can then implement Deref to make accessing these properties more natural, though we don't have to. But I think we should bite the bullet and accept having to repeat the same code in lots of Nodes, to keep the AST looking clean and understandable.

krobelus commented:
> We can probably eliminate the const to make it object safe by using a const function instead of a constant.

Object safe yes but we cannot call generic methods on `dyn Node`

```rust
pub trait Node {
    fn try_as<T: Node>(&self) -> Option<&T> {
        todo!()
    }
}

pub trait NodeVisitor<'a> {
    fn visit<T: Node>(&mut self, node: &T);
}

pub struct Traversal<'a> {
    stack: Vec<&'a dyn Node>,
}

impl<'a> NodeVisitor<'a> for Traversal<'a> {
    fn visit(&mut self, node: &dyn Node) {}
}
```

ridiculousfish commented:
Ugh that limitation on no-generics-in-dyn sucks. There's got to be some sort of best practice for expressing ASTs in Rust.

Xiretza commented:
You can have generic methods on `dyn` if you implement them on the `dyn` itself, something like this:

```rust
impl dyn Node {
    pub fn try_as<T: Node>(&self) -> Option<&T> {
        todo!()
    }
}
```

mqudsi commented:
Can I again ask why we can't use an enum for this? For only a handful of node types, it's the logical approach. You'll centralize the handling of variants. You might not need to use dynamic dispatch at all anymore, which means everything can be on the stack.

If you need to implement traits for each variant and don't want to do it by hand, you can use [the enum_dispatch crate](https://crates.io/crates/enum_dispatch). But I don't imagine that'll be a big problem here.

ridiculousfish commented:
How would an enum work? For example `if_statement_t` has its `if` clause, a list of `elseif` clauses, the `else` clause, the `end` keyword, and a list of redirections. We want all of those fields to be concrete types, not a node which can be any type.

mqudsi commented:
What I've seen is separate top-level types for all node specializations, but an enum used only to take the place of the `&dyn Node`.

So you would have something like this:

```rust
enum Node {
   If(IfNode),
   ElseIf(ElseIfNode),
   EndIf(EndIfNode),
   Begin(BeginNode),
   ...
}
```

where each of those is its own type. Then you can either implement top-level behavior directly in the `impl Node {}` block or by implementing a trait and dispatching to that trait in the same impl but for `Node`.

e.g. option 1

```rust
impl Node {
    fn foo(&self) {
        match self {
            Node::IfNode => // do something specific to if,
            Node::ElseIfNode => // do something specific to else if,
            // etc
        }
    }
}
```

or option 2

```rust
trait Foo { fn foo(&self); }

impl Foo for IfNode {}
impl Foo for ElseNode {}

impl Foo for Node {
    fn foo(&self) {
        match self {
            Node::IfNode(node) => node.foo(),
            Node::ElseNode(node) => node.foo(),
       }
}
```

the latter impl of which the enum_dispatch trait automates for you if all nodes implement Foo.

ridiculousfish commented:
It strikes me that we don't need to make `try_as` a member function, it may as well be a free function, because it's not generic.

Xiretza commented:
Personally I agree with @mqudsi that enums would be the more convenient and ergonomic approach. Dynamic dispatch is most useful if you don't control all the types that implement `Node` - we're not a library, so we do. Another concern might be that an enum is as big as its largest member; if that's a problem, the enum can contain `Box`es instead of the objects themselves, making it pointer-sized.

krobelus commented on fish-rust/src/common.rs:
```diff
@@ -96,6 +158,544 @@ pub fn escape_string(s: &wstr, style: EscapeStringStyle) -> WString {
+                        }
+                    } else {
+                        result_char_or_none =
+                            Some(unsafe { char::from_u32_unchecked(res.try_into().unwrap()) });
```
> One thing we haven't thought about is what happens if the user tries to encode a surrogate code point, which Rust disallows; for example \UDE01. Although, this doesn't seem to do anything today.

I'll replace it with `�` (`\UFFFD`, replacement character) for now. We could also preserve the codepoint using the private use area but that doesn't tell us how to print it. I guess printing � is better than silently dropping it.

krobelus commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+    fn visit(&mut self, node: &'a dyn Node);
+}
+
+pub trait Acceptor {
```
Acceptor needs to be `pub` because `Traversal::new` takes a node that we need to call `accept()` on

krobelus commented on fish-rust/src/ast.rs:
```diff
@@ -0,0 +1,5206 @@
+
+/// Trait for all "leaf" nodes: nodes with no ast children.
+pub trait Leaf: Node {
+    fn as_node(&self) -> &dyn Node;
```
> I think we should be able to provide a default implementation of as_node here,

No, it says "trait upcasting is experimental"

krobelus commented:
I implemented the syntactic changes.

Even if we implement the generic `try_as` we still need the concrete methods like `as_argument()` to do the actual conversion so we'd add double the number of symbols to the object file.

The enum approach is sort of unrelated but it sounds interesting, it will remove some of the dynamic dispatching I added.

The added copying at the FFI boundary makes things slower (memcpy dominates the profile) but it's not unusable, which is good news:

	$ hyperfine ./fish.{old,new}" -c 'source ../share/completions/git.fish'"
	Benchmark 1: ./fish.old -c 'source ../share/completions/git.fish'
	  Time (mean ± σ):     195.5 ms ±   2.9 ms    [User: 190.1 ms, System: 4.4 ms]
	  Range (min … max):   193.2 ms … 205.1 ms    15 runs
	 
	Benchmark 2: ./fish.new -c 'source ../share/completions/git.fish'
	  Time (mean ± σ):     677.5 ms ±  62.0 ms    [User: 665.4 ms, System: 10.0 ms]
	  Range (min … max):   611.7 ms … 805.5 ms    10 runs
	 
	Summary
	  './fish.old -c 'source ../share/completions/git.fish'' ran
	    3.47 ± 0.32 times faster than './fish.new -c 'source ../share/completions/git.fish''

mqudsi commented:
Continuing the discussion from matrix, if I get some time I'd be happy to take a look at switching this to use the enum approach and seeing how that unfolds.

krobelus commented:
Now that we no longer need the FFI we could do

```
enum Node {
    Leaf(Leaf),
    Branch(Branch),
    List(List),
}
enum Leaf {
    String(String),
    SemiNl(SemiNl),
    ...
}
```

```
diff --git a/src/bin/fish_indent.rs b/src/bin/fish_indent.rs
index 711dca1d0..c80ea07d5 100644
--- a/src/bin/fish_indent.rs
+++ b/src/bin/fish_indent.rs
@@ -302,7 +302,7 @@ impl<'source, 'ast> PrettyPrinterState<'source, 'ast> {
                         let p = p.parent().unwrap();
                         assert_eq!(p.typ(), Type::statement);
                         let p = p.parent().unwrap();
-                        if let Some(job) = p.as_job_pipeline() {
+                        if let Node::Branch(Branch::JobPipeline(job)) = p {
                             if !job.variables.is_empty() {
                                 result.allow_escaped_newlines = true;
                             }
```

and use the `enum_dispatch` crate.

mqudsi commented:
Future AST suggestion: https://www.cs.cornell.edu/~asampson/blog/flattening.html
